icu4c/source/i18n/ucal.cpp

```diff
@@ -167,21 +168,15 @@ ucal_open(  const char16_t*  zoneID,
   }
 
   if ( caltype == UCAL_GREGORIAN ) {
-      char localeBuf[ULOC_LOCALE_IDENTIFIER_CAPACITY];
       if ( locale == nullptr ) {
           locale = uloc_getDefault();
       }
-      int32_t localeLength = static_cast<int32_t>(uprv_strlen(locale));
-      if (localeLength >= ULOC_LOCALE_IDENTIFIER_CAPACITY) {
-          *status = U_ILLEGAL_ARGUMENT_ERROR;
-          return nullptr;
-      }
-      uprv_strcpy(localeBuf, locale);
-      uloc_setKeywordValue("calendar", "gregorian", localeBuf, ULOC_LOCALE_IDENTIFIER_CAPACITY, status);
+      CharString localeBuf(locale, *status);
+      ulocimp_setKeywordValue("calendar", "gregorian", localeBuf, status);
       if (U_FAILURE(*status)) {
           return nullptr;
       }
-      return (UCalendar*)Calendar::createInstance(zone.orphan(), Locale(localeBuf), *status);
+      return (UCalendar*)Calendar::createInstance(zone.orphan(), Locale(localeBuf.data()), *status);
```

Comment:
Since we are eventually constructing a C++ Locale, wouldn't it be easier to do that first and then call its .setKeywordValue()?

Reply:
That sounds like a good idea, but let's make one change at a time, there's a nearly endless amount of improvements that could be done in this code base and we have bigger fish to fry.